modules/peers.js

```diff
@@ -304,139 +383,228 @@ Peers.prototype.acceptable = function (peers) {
+	options.attempts = ['matched broadhash', 'unmatched broadhash'];
 	options.attempt = 0;
 	options.matched = 0;
 
-	if (!options.broadhash) {
-		delete options.broadhash;
-	}
-
 	function randomList (options, peers, cb) {
-		library.db.query(sql.randomList(options), options).then(function (rows) {
-			options.limit -= rows.length;
-			if (options.attempt === 0 && rows.length > 0) { options.matched = rows.length; }
-			library.logger.debug(['Listing', rows.length, options.attempts[options.attempt], 'peers'].join(' '));
-			return setImmediate(cb, null, self.acceptable(peers.concat(rows)));
-		}).catch(function (err) {
-			library.logger.error(err.stack);
-			return setImmediate(cb, 'Peers#list error');
+		// Get full peers list (random)
+		__private.getByFilter ({}, function (err, peersList) {
+			var accepted, found, matched, picked;
+
+			found = peersList.length;
+			// Apply filters
+			peersList = peersList.filter(function (peer) {
+				if (options.broadhash) {
+					// Skip banned peers (state 0)
+					return peer.state > 0 && (
+						// Matched broadhash when attempt 0
+						options.attempt === 0 ? (peer.broadhash === options.broadhash) : 
+						// Unmatched broadhash when attempt 1
+						options.attempt === 1 ? (peer.broadhash !== options.broadhash) : false
+					);
+				} else {
+					// Skip banned peers (state 0)
+					return peer.state > 0;
+				}
+			});
+			matched = peersList.length;
+			// Apply limit
+			peersList = peersList.slice(0, options.limit);
+			picked = peersList.length;
+			accepted = self.acceptable(peers.concat(peersList));
+			library.logger.debug('Listing peers', {attempt: options.attempts[options.attempt], found: found, matched: matched, picked: picked, accepted: accepted.length});
+			return setImmediate(cb, null, accepted);
 		});
 	}
 
-	async.waterfall([
-		// Matched broadhash
+	steed.waterfall([
 		function (waterCb) {
-			return randomList(options, [], waterCb);
+			// Matched broadhash
+			return randomList (options, [], waterCb);
 		},
-		// Unmatched broadhash
 		function (peers, waterCb) {
+			options.matched = peers.length;
+			options.limit -= peers.length;
+			++options.attempt;
 			if (options.limit > 0) {
-				options.attempt += 1;
-
-				return randomList(options, peers, waterCb);
-			} else {
-				return setImmediate(waterCb, null, peers);
-			}
-		},
-		// Fallback
-		function (peers, waterCb) {
-			delete options.broadhash;
-
-			if (options.limit > 0) {
-				options.attempt += 1;
-
+				// Unmatched broadhash
 				return randomList(options, peers, waterCb);
 			} else {
 				return setImmediate(waterCb, null, peers);
 			}
 		}
 	], function (err, peers) {
+		// Calculate consensus
 		var consensus = Math.round(options.matched / peers.length * 100 * 1e2) / 1e2;
 		    consensus = isNaN(consensus) ? 0 : consensus;
 
-		library.logger.debug(['Listing', peers.length, 'total peers'].join(' '));
+		library.logger.debug('Listing ' + peers.length + ' total peers');
```

Comment:
Can we stick to joining array for now? This standard set throughout. We can change all separately if deemed better.